README.md
# viterbi-algorithm-clj

西瓜书和PRML学习记录--HMM实现算法,以及数学公式理论到算法的过程记录

## Usage

```clojure

viterbi-algorithm-clj.core=> (Viterbi/main (into-array (list "1" "3")))

States: #, NN, VB,

Observations: I, write, a letter,

Start probability: #: 0.3, NN: 0.4, VB: 0.3,

Transition probability:
 #: {  #: 0.2,   NN: 0.2,   VB: 0.6, }
 NN: {  #: 0.4,   NN: 0.1,   VB: 0.5, }
 VB: {  #: 0.1,   NN: 0.8,   VB: 0.1, }


Emission probability:
 #: {  I: 0.01,   write: 0.02,   a letter: 0.02, }
 NN: {  I: 0.8,   write: 0.01,   a letter: 0.5, }
 VB: {  I: 0.19,   write: 0.97,   a letter: 0.48, }
Viterbi path: [NN, VB, NN, ].
 Probability of the whole system: 0.06208000000000002
nil
viterbi-algorithm-clj.core=>

```

## License

Copyright © 2017 FIXME

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2017-10-18
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2017-10-18
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/viterbi-algorithm-clj/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/viterbi-algorithm-clj/compare/0.1.0...0.1.1


test/viterbi_algorithm_clj/core_test.clj
(ns viterbi-algorithm-clj.core-test
  (:require [clojure.test :refer :all]
            [viterbi-algorithm-clj.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/viterbi_algorithm_clj/core.clj
(ns viterbi-algorithm-clj.core
  (:use [clojure.pprint])
  (:import (viterbiAlgorithm Viterbi)))

(defstruct hmm :n :m :init-probs :emission-probs :state-transitions)

(defn make-hmm [{:keys [states, obs, init-probs, emission-probs, state-transitions]}]
  (struct-map hmm
              :n (count states)
              :m (count obs)
              :states states
              :obs obs
              :init-probs init-probs ;; n dim
              :emission-probs emission-probs ;;m x n
              :state-transitions state-transitions))

(defn indexed [s]
  (map vector (iterate inc 0) s))

(defn argmax [coll]
  (loop [s (indexed coll)
         max (first s)]
    (if (empty? s)
      max
      (let [[idx elt] (first s)
            [max-indx max-elt] max]
        (if (> elt max-elt)
          (recur (rest s) (first s))
	  (recur (rest s) max))))))

(defn pprint-hmm [hmm]
  (println "number of states: " (:n hmm) " number of observations:  " (:m hmm))
  (print "init probabilities: ") (pprint (:init-probs hmm))
  (print "emission probs: " ) (pprint (:emission-probs hmm))
  (print "state-transitions: " ) (pprint (:state-transitions hmm)))

(defn init-alphas [hmm obs]
  (map (fn [x]
         (* (aget (:init-probs hmm) x) (aget (:emission-probs hmm) x obs)))
       (range (:n hmm))))

(defn forward [hmm alphas obs]
  (map (fn [j]
         (* (reduce (fn [sum i]
                      (+ sum (* (nth alphas i) (aget (:state-transitions hmm) i j))))
                    0
                    (range (:n hmm)))
            (aget (:emission-probs hmm) j obs))) (range (:n hmm))))

(defn delta-max [hmm deltas obs]
  (map (fn [j]
         (* (apply max (map (fn [i]
                              (* (nth deltas i)
                                 (aget (:state-transitions hmm) i j)))
                            (range (:n hmm))))
            (aget (:emission-probs hmm) j obs)))
       (range (:n hmm))))

(defn backtrack [paths deltas]
  (loop [path (reverse paths)
         term (first (argmax deltas))
         backtrack []]
    (if (empty? path)
      (reverse (conj backtrack term))
      (recur (rest path) (nth (first path) term) (conj backtrack term)))))

(defn update-paths [hmm deltas]
  (map (fn [j]
         (first (argmax (map (fn [i]
                               (* (nth deltas i)
                                  (aget (:state-transitions hmm) i j)))
                             (range (:n hmm))))))
       (range (:n hmm))))

(defn viterbi [hmm observs]
  (loop [obs (rest observs)
         alphas (init-alphas hmm (first observs))
         deltas alphas
         paths []]
    (if (empty? obs)
      [(backtrack paths deltas) (float (reduce + alphas))]
      (recur (rest obs)
             (forward hmm alphas (first obs))
             (delta-max hmm deltas (first obs))
             (conj paths (update-paths hmm deltas))))))


src/java/Viterbi.java
package viterbiAlgorithm;

public class Viterbi {
	public static String[] states = { "#", "NN", "VB" };
	public static String[] observations = { "I", "write", "a letter" };
	public static double[] start_probability = { 0.3, 0.4, 0.3 };
	public static double[][] transition_probability = { { 0.2, 0.2, 0.6 },
			{ 0.4, 0.1, 0.5 }, { 0.1, 0.8, 0.1 } };
	public static double[][] emission_probability = { { 0.01, 0.02, 0.02 },
			{ 0.8, 0.01, 0.5 }, { 0.19, 0.97, 0.48 } };
	public static class TNode {
		public int[] v_path;
		public double v_prob;

		public TNode( int[] v_path, double v_prob) {
			this.v_path = copyIntArray(v_path);
			this.v_prob = v_prob;
		}
	}

	public static int[] copyIntArray(int[] ia) {
		int[] newIa = new int[ia.length];
		for (int i = 0; i < ia.length; i++) {
			newIa[i] = ia[i];
		}
		return newIa;
	}

	public static int[] copyIntArray(int[] ia, int newInt) {
		int[] newIa = new int[ia.length + 1];
		for (int i = 0; i < ia.length; i++) {
			newIa[i] = ia[i];
		}
		newIa[ia.length] = newInt;
		return newIa;
	}

	// forwardViterbi(observations, states, start_probability,
	// transition_probability, emission_probability)
	public int[] forwardViterbi(String[] y, String[] X, double[] sp,
			double[][] tp, double[][] ep) {
		TNode[] T = new TNode[X.length];
		for (int state = 0; state < X.length; state++) {
			int[] intArray = new int[1];
			intArray[0] = state;
			T[state] = new TNode( intArray, sp[state] * ep[state][0]);
		}

		for (int output = 1; output < y.length; output++) {
			TNode[] U = new TNode[X.length];
			for (int next_state = 0; next_state < X.length; next_state++) {
				int[] argmax = new int[0];
				double valmax = 0;
				for (int state = 0; state < X.length; state++) {
					int[] v_path = copyIntArray(T[state].v_path);
					double v_prob = T[state].v_prob;
					double p = ep[next_state][output] * tp[state][next_state];
					v_prob *= p;
					if (v_prob > valmax) {
						if (v_path.length == y.length) {
							argmax = copyIntArray(v_path);
						} else {
							argmax = copyIntArray(v_path, next_state);
						}
						valmax = v_prob;

					}
				}
				U[next_state] = new TNode( argmax, valmax);
			}
			T = U;
		}
		// apply sum/max to the final states:
		int[] argmax = new int[0];
		double valmax = 0;
		for (int state = 0; state < X.length; state++) {
			int[] v_path = copyIntArray(T[state].v_path);
			double v_prob = T[state].v_prob;
			if (v_prob > valmax) {
				argmax = copyIntArray(v_path);
				valmax = v_prob;
			}
		}
		System.out.print("Viterbi path: [");
		for (int i = 0; i < argmax.length; i++) {
			System.out.print(states[argmax[i]] + ", ");
		}
		System.out.println("].\n Probability of the whole system: " + valmax);
		return argmax;
	}

	public static void main(String[] args) throws Exception {
		System.out.print("\nStates: ");
		for (int i = 0; i < states.length; i++) {
			System.out.print(states[i] + ", ");
		}
		System.out.print("\n\nObservations: ");
		for (int i = 0; i < observations.length; i++) {
			System.out.print(observations[i] + ", ");
		}
		System.out.print("\n\nStart probability: ");
		for (int i = 0; i < states.length; i++) {
			System.out.print(states[i] + ": " + start_probability[i] + ", ");
		}
		System.out.println("\n\nTransition probability:");
		for (int i = 0; i < states.length; i++) {
			System.out.print(" " + states[i] + ": {");
			for (int j = 0; j < states.length; j++) {
				System.out.print("  " + states[j] + ": "
						+ transition_probability[i][j] + ", ");
			}
			System.out.println("}");
		}
		System.out.println("\n\nEmission probability:");
		for (int i = 0; i < states.length; i++) {
			System.out.print(" " + states[i] + ": {");
			for (int j = 0; j < observations.length; j++) {
				System.out.print("  " + observations[j] + ": "
						+ emission_probability[i][j] + ", ");
			}
			System.out.println("}");
		}
		Viterbi v=new Viterbi();
		v.forwardViterbi(observations, states, start_probability,
				transition_probability, emission_probability);
		
	}
}
